### PR TITLE
Interpreter: add InstructionParts and decode it

### PR DESCRIPTION
**Pull request by dannywillems**
See https://max.cs.kzoo.edu/cs230/Resources/MIPS/MachineXL/InstructionFormats.html

For the instruction I and J, we will simply add the relevant parts.
The next PR will be the jump instruction.

### CODE REVIEW

File: optimism/src/mips/witness.rs [hunk 1]
```diff
@@ -361,7 +379,7 @@ impl<Fp: Field> Env<Fp> {
                 .unwrap_or_else(|| "n/a".to_string());
 
             info!(
-                "processing step {} pc {:#010x} insn {:#010x} ips {:.2} page {} mem {} name {}",
+                "processing step={} pc={:#010x} insn={:#010x} ips={:.2} page={} mem={} name={}",
```
Comment (rbonichon) [inline optimism/src/mips/witness.rs:382]: The formatting here was the one that `cannon` uses, in order to ease the comparison between the outputs of the 2 VMs. Has `cannon`'s formatting changed?

File: optimism/src/mips/interpreter.rs [hunk 2]
```diff
@@ -8,6 +10,42 @@ pub const FD_HINT_WRITE: u32 = 4;
...
+    fn index(&self, index: InstructionPart) -> &Self::Output {
+        match index {
+            InstructionPart::OpCode => &self.op_code,
+            InstructionPart::RS => &self.rs,
+            InstructionPart::RT => &self.rt,
+            InstructionPart::RD => &self.rd,
+            InstructionPart::Shamt => &self.shamt,
+            InstructionPart::Funct => &self.funct,
+        }
+    }
```
[the remark this answers is not in the document]
Comment (dannywillems) [reply optimism/src/mips/interpreter.rs:46]: Interesting usecase: https://github.com/o1-labs/proof-systems/pull/1361/files#diff-70c1a6af8b08e80764e14c1efef06d463db55758eada514596b298d34fdcf2f2R217

**Review by querolita — APPROVED** [review]
Left a few questions and suggestions, but looks okay for me to merge.

File: optimism/src/mips/interpreter.rs [hunk 3]
```diff
@@ -8,6 +10,42 @@ pub const FD_HINT_WRITE: u32 = 4;
...
+    OpCode,
+    RS,
+    RT,
+    RD,
+    Shamt,
+    Funct,
+}
+
+#[derive(Debug, Clone, Copy, Eq, PartialEq, Default, Serialize, Deserialize)]
+pub struct InstructionParts<T> {
```
[the remark this answers is not in the document]
Comment (dannywillems) [reply optimism/src/mips/interpreter.rs:24]: Yes they do. The `op_code` is always 6 bits. We can add and shifts for the others.

File: optimism/src/mips/witness.rs [hunk 4]
```diff
@@ -176,122 +181,135 @@ impl<Fp: Field> Env<Fp> {
         panic!("Could not access address")
     }
 
-    pub fn decode_instruction(&self) -> Instruction {
+    pub fn decode_instruction(&self) -> (Instruction, u32) {
```
[the remark this answers is not in the document]
Comment (dannywillems) [reply optimism/src/mips/witness.rs:184]: I keep a follow-up issue for this: https://github.com/o1-labs/proof-systems/issues/1363

File: optimism/src/mips/witness.rs [hunk 5]
```diff
@@ -176,122 +181,135 @@ impl<Fp: Field> Env<Fp> {
...
+        };
+        (opcode, instruction)
     }
 
     pub fn step(&mut self, config: VmConfiguration, metadata: &Meta, start: &Start) {
-        let instruction = self.decode_instruction();
-        println!("instruction: {:?}", instruction);
+        let (opcode, instruction) = self.decode_instruction();
+        let instruction_parts: InstructionParts<u32> = InstructionParts {
+            op_code: ((instruction >> 26) & ((1 << (32 - 26)) - 1)),
```
[the remark this answers is not in the document]
Comment (dannywillems) [reply optimism/src/mips/witness.rs:303]: Yes, it is the same.

**Pull request merged**